source/common/router/router.cc

```diff
@@ -151,13 +152,39 @@ FilterUtility::finalTimeout(const RouteEntry& route, Http::HeaderMap& request_he
   }
   timeout.per_try_timeout_ = route.retryPolicy().perTryTimeout();
 
-  Http::HeaderEntry* header_timeout_entry = request_headers.EnvoyUpstreamRequestTimeoutMs();
   uint64_t header_timeout;
-  if (header_timeout_entry) {
-    if (absl::SimpleAtoi(header_timeout_entry->value().getStringView(), &header_timeout)) {
-      timeout.global_timeout_ = std::chrono::milliseconds(header_timeout);
+
+  if (respect_expected_rq_timeout) {
+    // Check if there is timeout set by egress envoy.
+    // If present, use that value as route timeout and don't override
+    // `x-envoy-expected-rq-timeout-ms` header. At this point `x-envoy-upstream-rq-timeout-ms`
+    // header should have been sanitized by egress envoy.
+    Http::HeaderEntry* header_expected_timeout_entry =
+        request_headers.EnvoyExpectedRequestTimeoutMs();
+    if (header_expected_timeout_entry) {
+      // This will prevent from overriding `x-envoy-expected-rq-timeout-ms` header.
```

Comment:
@snowp this was protection against this code path:
```
  if (insert_envoy_expected_request_timeout_ms && expected_timeout > 0) {
    request_headers.insertEnvoyExpectedRequestTimeoutMs().value(expected_timeout);
  }
```
We do indeed derive timeout to a separate data structure (with global timeout), so there is not a big gain from setting `insert_envoy_expected_request_timeout_ms` to false.

Reply:
changed my mind, there is a gain, so that we use same value in timeout.global_timeout (derived from `x-envoy-expected-timeout-ms`) and observe same value in header `x-envoy-expected-timeout-ms` by not overriding it.

Reply:
Hmm, I think we want the expected header to reflect the timeout used by the router, which is affected by more than just those two headers. If you look further down in that function you'll see that we'll use the per try timeout instead of the global timeout if it's set. It seems like we want the expected timeout header to always reflect the timeout enforced by the router for the outgoing request, and just use the incoming expected timeout header to infer the *global* timeout. Does that make sense?

Reply:
@snowp it does, thanks for clarification

Reply:
will adjust the test as well